src/plugins_3rdparty/primer3/src/Primer3Task.cpp

```diff
@@ -781,39 +743,89 @@ QString Primer3ToAnnotationsTask::generateReport() const {
                .arg(pairStats.compl_end)
                .arg(pairStats.ok);
 
+    if (checkComplementTask != nullptr) {
+        res += checkComplementTask->generateReport();
+    }
+
     return res;
 }
 
-Task::ReportResult Primer3ToAnnotationsTask::report() {
-    if (hasError() || isCanceled()) {
-        return ReportResult_Finished;
+
+void Primer3ToAnnotationsTask::findExonTaskIsfinished() {
+    QList<U2Region> regions = findExonsTask->getRegions();
+    if (regions.isEmpty()) {
+        setError(tr("Failed to find any exon annotations associated with the sequence %1."
+            "Make sure the provided sequence is cDNA and has exonic structure annotated")
+            .arg(seqObj->getSequenceName()));
+        return;
+    } else {
+        const U2Range<int>& exonRange = settings->getSpanIntronExonBoundarySettings().exonRange;
+
+        if (exonRange.minValue != 0 && exonRange.maxValue != 0) {
+            int firstExonIdx = exonRange.minValue;
+            int lastExonIdx = exonRange.maxValue;
+            if (firstExonIdx > regions.size()) {
+                setError(tr("The first exon from the selected range [%1,%2] is larger the number of exons (%3)."
+                    " Please set correct exon range.")
+                    .arg(firstExonIdx)
+                    .arg(lastExonIdx)
+                    .arg(regions.size()));
+                return;
+            }
+
+            if (lastExonIdx > regions.size()) {
+                setError(tr("The the selected exon range [%1,%2] is larger the number of exons (%3)."
+                    " Please set correct exon range.")
+                    .arg(firstExonIdx)
+                    .arg(lastExonIdx)
+                    .arg(regions.size()));
+                return;
+            }
+
+            regions = regions.mid(firstExonIdx - 1, lastExonIdx - firstExonIdx + 1);
+            int totalLen = 0;
+            for (const U2Region& r : regions) {
+                totalLen += r.length;
+            }
+            settings->setIncludedRegion(regions.first().startPos + settings->getFirstBaseIndex(), totalLen);
+        }
+        settings->setExonRegions(regions);
+        // reset target and excluded regions regions
+        QList<U2Region> emptyList;
+        settings->setExcludedRegion(emptyList);
+        settings->setTarget(emptyList);
     }
-    CHECK_EXT(!annotationTableObject.isNull(), setError(tr("Object with annotations was removed")), ReportResult_Finished);
-    SAFE_POINT(searchTask != nullptr, L10N::nullPointerError("Primer3Task"), ReportResult_Finished);
+}
 
-    const QList<PrimerPair>& bestPairs = searchTask->getBestPairs();
+QMap<QString, QList<SharedAnnotationData>> Primer3ToAnnotationsTask::getResultAnnotations() const {
+    QList<QSharedPointer<PrimerPair>> filteredPrimers;
+    if (checkComplementTask != nullptr) {
+        filteredPrimers = checkComplementTask->getFilteredPrimers();
+    }
+    const auto& bestPairs = primer3Task->getBestPairs();
```

Comment:
primer3Task is nullable. Add a CHECK or SAFE_POINT before use

Reply:
`SAFE_POINT` would be better, `primer3Task` shouldn't be `nullptr` here